### ISSUE TITLE
`translateProps` function feature proposal

### ISSUE THREAD

**Issue by souldreamer** [Sawtaytoes/OneForm#13]
Was wondering if it wouldn't help to send the default computed child props to the translateProps function as one of the parameters. Most of the time you only need to change the behavior of one or two of the child properties, so copy/pasting the entire code to generate the properties seems wasteful.

The proposed code change would be in `Field.jsx`:

```typescript
const DEFAULT_TRANSLATE_PROPS = ({defaultProps}) => defaultProps;

const Field = ({
  children,
  isCheckboxElement = false,
  isMultipleElement = false,
  translateProps = (
    DEFAULT_TRANSLATE_PROPS
  ),
}) => {
```

```typescript
  const childProps = useMemo(
    () =>
      translateProps({
        errorMessages,
        fieldName,
        isChecked,
        isVisited,
        updateFieldValue,
        value,
        visitField,
        defaultProps: isHtmlElement
          ? {
              checked: isChecked,
              'data-error': errorMessages.length > 0 ? true : null,
              'data-visited': isVisited ? true : null,
              name: fieldName,
              onBlur: visitField,
              onChange: updateFieldValue,
              value,
            }
          : {
              checked: isChecked,
              dirty: isVisited,
              error: Boolean(errorMessages[0]),
              errorMessages: errorMessages,
              errors: errorMessages,
              isChecked,
              isDirty: isVisited,
              isTouched: isVisited,
              isVisited,
              name: fieldName,
              onBlur: visitField,
              onChange: updateFieldValue,
              touched: isVisited,
              value,
              visited: isVisited,
            },
      }),
    [errorMessages, fieldName, isChecked, isHtmlElement, isVisited, translateProps, updateFieldValue, value, visitField],
  );
```

A lot of the time you just need to change the behavior of `onBlur` and/or `onChange` and want to keep the other properties the same, and rewriting the same code is very non-DRY.

Another possibility is to do something like `childProps = {...defaultProps, ...translateProps(...)}` which would probably be even DRY-er, but at the cost of still passing on the properties the user might not want to pass, so that one would probably be a case of being *too* DRY.

----

While we're at it, passing the `inputValue`, `isCheckbox`, and `isMultiple` (though the latter two *might* be okay to be inferred because you'd be using a different translate function for them, but that's not always necessary, so having the info in the function would probably help in a few cases, and would definitely not hinder anything) into the `translateProps` function would probably not be amiss either (nor would passing all the other child props along), so we don't have to replicate logic from the `<Field />` component and just use `<Field />`s with `translateProps` more.

While building a custom component isn't bad, re-using the `<Field />` component should be preferred IMHO.

Also note that the current `<MaterialUiField />` component nukes the `helperText` the user might set on the child component when there's no error. Its `translateProps` function should do something like `helperText: errorMessages[0] || childProps.helperText`.

Not to mention that the library having a `<MaterialUiField />` is a bit redundant in its current form, at most you should have a `materialUiTranslateProps` export, or even better a tutorial in the docs where the code is shown and explained.

----

Also, hope I'm not bugging you too much with feature/change requests 😅

**Comment by Sawtaytoes** [comment]
## Your own `Field`
Making your own `Field` component should be a one-time thing. With Formik and many other libraries, you have to wrap each component in their `Field` which makes it a pain to use and tightly couples your component library with your form library. OneForm's `Field` is a way to avoid tight coupling.

At work, we have a `CompanyField` component (not the actual name) using `translateProps` that works with every component in our component library except one super weird one.

For the one-off component, I wrote a `OneOffComponentField` with `useField`. I didn't use `translateProps` because I needed to add in a bunch of extra callbacks and state to get this particular component working properly. `OneOffComponentField` is still separate from `OneOffComponent`.

## Other Props
`translateProps` is designed to be a simple 1:1 props renamer.

`inputValue`, `isCheckbox`, and `isMultiple` are fine to pass, but the fact that you're touching `onBlur` and `onChange` is an issue. For situations where you need to create new values and callbacks, you should be using `useField`.

Everything in `translateProps` needs to be memoized, or you'll get unnecessary re-renders. I have a presentation on this very topic (unrelated to OneForm) at KCDC this week which I'll release to YouTube after the conference. It's a problem with orders of magnitude.

I can see why you'd want those props available, but doing it allows something other than a 1:1 props conversion.

## `MaterialUiField`
You're correct about the `MaterialUiField` example. I don't export it from OneForm though. Its existence is only there to test `translateProps`.

Do you think it should be exported? If so, I'd wanna put it in another library or figure out another way to handle it.

It's no big deal if people have tree-shaking enabled, but I've found projects at work using `oneform.umd.min.js` instead of the module files directly (possibly a config issue in those projects) meaning it would be included in a bundle when it's not used.

Another worry is Module Federation. You need to disable tree-shaking when using it, so I'd wanna slim down the bundle in that case. Not a huge deal, but something to keep in mind.

## Default props
_**tl;dr** I don't think adding `defaultProps` would be a good idea._

I actually don't like how `Field` guesses the child components' props. It's hacky and doesn't scale. I did it to avoid creating custom `Field` components with `useField` prior to the introduction of `translateProps`. Now that we have a good solution, `Field` should only output valid-HTML props. Changing this would break existing apps, but I should've done it from the start. I'm leaving this for a `v2` release.

As you said, doing automatic `...defaultProps` would be problematic. In a lot of cases, component libraries (like Material-UI) do `<input {...rest} />`. This causes a bunch of React HTML errors and is precisely why I use `isHtmlElement`.

Even if we wanted to include `defaultProps` as another object property, I want people to be explicit about what they expect their components to require. No one knows what's inside of `defaultProps` at any given OneForm version. It would make upgrades scary and make it difficult for me to make changes to `Field` over time. It also goes against why `translateProps` exists in the first place; to be a 1:1 props rename.

## Conclusion
You can still create your own `Field` component with a more-advanced `translateProps` using `useField`. That way you can get the functionality you want while leaving the default library memoization-safe.

**Comment by souldreamer** [comment]
You *do* export `MaterialUiField`. From the latest version of `index.js`:
```typescript
export { default as MaterialUiField } from './MaterialUiField.jsx'
```

I'm on the side of not exporting it, just explaining how it would work in the docs so people get an idea of how they can build on `Field` for custom components.

---

## `translateProps`

> Everything in translateProps needs to be memoized, or you'll get unnecessary re-renders

That's pretty much true of anything in React 😁. But don't see how adding the properties would change anything about memoization:
* `inputValue` - this isn't necessary anymore now that we're also checking inputValue in `useField` when railroading it with a `[]`
* `isCheckbox` and `isMultiple` - would allow writing only one translateProps function and re-using it for most components (they don't change that often, usually only once when creating the `Field`, so no memo/rerendering issues here)
* passing in the props that `Field` would use if you didn't return anything from `translateProps` (or didn't have it) (what I called `defaultProps`) - already change based on the same things `translateProps` is called for

`translateProps` would return a mapping still, it would just have a bit more access to information so it can create that mapping, and it would also have the benefit of not needing to create the entire mapping when only one thing is different. For example, a component could have an `onChange` that returns the actual value, but also wants to visit the field. Currently, I'd have to copy/paste the entire object mapping from `Field.jsx` for this and just make the onChange be 
```typescript
// ...go to your source code and copy/paste stuff...
onChange: (e) => { visitField(); updateFieldValue(e); }
````

Whereas with stuff like passing in the `defaultComputedMapping` (changed the name because `defaultProps` might suggest something else), I could do:
```typescript
{
  ...defaultComputedMapping,
  onChange: (e) => {...}
}
```
which would be a much better dev experience. Especially for more junior devs who really won't go into the library's source code.

My opinion is that if with very slight changes the `Field` component can do it with just having a global (in user-land, not in the library) `translateProps` function that we pass in (maybe even more for special cases), we should, and not just jump to recommending rewriting its logic in a custom `MyField` for just a small needed change. Especially when the mechanism is already there.

Don't get me wrong, I'm not afraid to get my own hands dirty (to make my multiple checkbox component work if/until the changes are made, I *did* write a custom Field. However, to gain more traction with users, the simpler you make these special cases (but not *super* special cases), the better.

For example (ignoring the fact that `useField` killed the `inputValue` for `isMultiple` (before my PR)), this is the multiple checkbox field component:

```typescript
function CheckboxField({
	children,
	isCheckboxElement,
	isMultipleElement,
}: {
	children: ReactElement;
	isMultipleElement?: boolean;
}) {
	const { value: inputValue, name } = Children.only(children).props;
	const { value: fieldDataValue } = useFieldData({ name });
	const checkboxFieldTranslate = useCallback(
		({ errorMessages, fieldName, isChecked, isVisited, updateFieldValue, value, visitField }: TranslateProps) => {
			return {
				checked: Array.isArray(fieldDataValue) ? fieldDataValue.includes(inputValue) : isChecked,
				dirty: isVisited,
				error: Boolean(errorMessages[0]),
				errorMessages: errorMessages,
				errors: errorMessages,
				isChecked,
				isDirty: isVisited,
				isTouched: isVisited,
				isVisited,
				name: fieldName,
				onBlur: visitField,
				onChange: updateFieldValue,
				touched: isVisited,
				value,
				visited: isVisited,
			};
		},
		[inputValue, fieldDataValue],
	);
	return (
		<Field isCheckboxElement isMultipleElement={isMultipleElement} translateProps={checkboxFieldTranslate}>
			{children}
		</Field>
	);
}

// ... when using:
<CheckboxField isMultipleElement><Checkbox></CheckboxField>
```

And this is what it could have been:
```typescript
/* global function, useCallback doesn't matter in this case */
const multiCheckboxTranslateProps = ({ value, inputValue, isChecked, defaultComputedProps }: TranslateProps) => ({ ...defaultComputedProps, checked: Array.isArray(value) ? value.includes(inputValue) : isChecked });

// ...when using:
<Field isCheckboxElement isMultipleElement translateProps={multiCheckboxTranslateProps}><Checkbox></Field>
```

IMHO it's more compact, more maintainable, and less mental clutter keeping track of things.

**Comment by Sawtaytoes** [comment]
## `MaterialUiField`
I didn't think I'd exported `MaterialUiField` since it wasn't in the docs, but it's in the docs too! My bad. Should've checked those first.

It's pretty clear I've forgotten a lot of the decisions made in the library 😔.

## `translateProps`
I also forgot how `translateProps` works. It memoizes the output in `useMemo` prior to passing those props into the cloned child component. Good catch.

In some ways, those props are memorized. The only time they're not is when some other prop changes. The way React renders, it doesn't actually care which prop changes. If one prop change occurs, the entire DOM output of that component is rewritten. So inline callbacks in `translateProps` is fine.

I think I understand why I used `useField` at work. I needed state, and `translateProps` doesn't have state. I originally tried hacking in a `useRef`, and that worked, but wasn't a great solution.

There's another optimization I could make. Right now, passing a new `translateProps` function gives you new values. Typically, you'd only want `translateProps` to execute when values in `Field` change. To that end, we'll want to create `translatePropsRef` and call `translatePropsRef.current()` instead.

With this change, you wouldn't need to touch `useField` at all.

## Your use case
Took me a while to figure out why you're calling `visitField` in `onChange`. You want a special case, for non-checkboxes where the error displays immediate rather than on blur. This could be useful for a password field; although, you could achieve something similar by returning a value in `onChange` that displays underneath the password field.

I can see now why there are more use cases for `translateProps` other than my original assumptions. We should definitely add `isCheckbox` and `isMultiple` in there.

I agree with your reasoning now for `defaultComputedMapping`; although, I'm wondering why we don't call it `defaultComputedProps` instead since it's the actual props that go into the child component by default.

I'm thinking this change will be for OneForm `v2`. Now that we have your other changes merged, I'll get to releasing `v1.9.0` and you can add those changes to another PR. While you're at it, we should also remove the `isHtmlElement` check and the props changes related to it.
